### PR TITLE
Enable GPGMM for DML backend.

### PR DESCRIPTION
**Pull request by bbernhar**
GPGMM introduces fast D3D12-based resource sub-allocation and video memory residency management for WebNN's DML backend.

### CODE REVIEW

File: build_overrides/gpgmm.gni [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# These are variables that are overridden by projects that include GPGMM.
+# The values in this file are the defaults for when we are building GPGMM
+# from WebNN-native's repository.
+
+# The paths to GPGMM's dependencies
+# Note: Non-standalone GPGMM has no additional dependencies.
```
Comment (huningxin) [inline build_overrides/gpgmm.gni:20]: Only comments in this file?

**Review by huningxin — COMMENTED** [review]
Thanks much @bbernhar . I have a high level question:

As you know, the device code of DML backend is based on upstream [DirectML device wrapper](https://github.com/microsoft/DirectML/blob/master/Python/src/device.h). The [plan](https://github.com/webmachinelearning/webnn-native/issues/3) is to reuse the upstream code in the future. So what's your plan to upstream gpgmm usage in upstream DirectML repo?

BTW, this repo use DEPS to manage dependency, so probably you don't need to check-in the git submodule of third_party/gpgmm.

File: build_overrides/gpgmm.gni (hunk 1 above)
Comment (huningxin) [reply build_overrides/gpgmm.gni]: makes sense. Thanks for the explanation.

**Review by huningxin — COMMENTED** [review]
> One DML project may wish to prioritize speed whereas another may care more about minimizing memory usage.

That's a good use case. Could you please share how to configure GPGMM regarding to speed and memory?

> GPGMM implements the same underlying D3D12 API used by the DirectML device wrapper (1:1) - meant to be easily integrated per DML project.

Sounds good.

As I understand GPGMM would also benefit Dawn, are there any pointers available to help understand the Dawn's usage of GPGMM?

File: src/webnn_native/dml/deps/src/device.h [hunk 2]
```diff
@@ -6,8 +6,17 @@
...
 #pragma once
 
+#include <gpgmm_d3d12.h>
+
 namespace pydml
 {
+    class SVDescriptorHeap : public gpgmm::d3d12::Heap {
+      public:
+        SVDescriptorHeap(ComPtr<ID3D12DescriptorHeap> heap,
+                                    uint64_t size);
```
[the remark this answers is not in the document]
Comment (bbernhar) [reply src/webnn_native/dml/deps/src/device.h]: Done.

**Comment by bbernhar** [comment]
> That's a good use case. Could you please share how to configure GPGMM regarding to speed and memory?

Most settings are documented [here, under `gpgmm::d3d12::ALLOCATOR_DESC ` ](https://github.com/intel/GPGMM/blob/d370f29db790da924ecf341c2a30b0342f497d07/src/d3d12/ResourceAllocatorD3D12.h#L53).

> are there any pointers available to help understand the Dawn's usage of GPGMM?

I would take a look at the default GPPM settings, as they are (currently) setup to be Dawn-equivalent. Eventually, (GPGMM) clients will be told to specify their own defaults. Dawn's defaults are very biased for fast small buffer uploads because at the time, we only had those types of workloads.

BTW, the DML team seems OK to integrate GPGMM. However, they use CMake (vs GN) so I'll plan for support in a future release - thanks for the suggestion!

**Comment by bbernhar** [comment]
@huningxin are we ready to merge or were you still reviewing? Note: the Node Binding bot failure looks unrelated to my change. Thanks!

**Review by huningxin — APPROVED** [review]
LGTM, thanks @bbernhar !

**Comment by BruceDai** [comment]
@fujunwei These two CI failure have been fixed by https://github.com/webmachinelearning/webnn-native/pull/159, please help rerun them, thanks.

**Pull request merged**